09-self-balancing-trees/AVL-tree/avlTree.c

```diff
@@ -0,0 +1,387 @@
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "avlTree.h"
+
+typedef struct Node {
+    NodeValue value;
+    Node* left;
+    Node* right;
+    short balance;
+} Node;
+
+Node* createTree(const char* key, const char* value, bool* errorCode) {
+    Node* root = calloc(1, sizeof(Node));
+    if (root == NULL) {
+        *errorCode = true;
+        return NULL;
+    }
+    root->value.key = key;
+    root->value.value = value;
+    return root;
+}
+
+void deleteTree(Node** root) {
+    if (*root == NULL) {
+        return;
+    }
+    deleteTree((&(*root)->left));
+    deleteTree((&(*root)->right));
+    free(*root);
+    *root = NULL;
+}
+
+Node* getLeftChild(Node* node, bool* errorCode) {
+    if (node == NULL) {
+        *errorCode = true;
+        return NULL;
+    }
+    return node->left;
+}
+
+Node* getRightChild(Node* node, bool* errorCode) {
+    if (node == NULL) {
+        *errorCode = true;
+        return NULL;
+    }
+    return node->right;
+}
+
+NodeValue getValue(Node* node, bool* errorCode) {
+    if (node == NULL) {
+        NodeValue incorrectNodeValues = { .key = -1, .value = NULL };
+        *errorCode = true;
+        return incorrectNodeValues;
+    }
+    return node->value;
+}
+
+const char* searchByKey(Node* node, const char* key) {
+    if (node == NULL) {
+        return NULL;
+    }
+
+    const char* foundValue = NULL;
+    if (strcmp(node->value.key, key) == 0) {
+        return node->value.value;
+    }
+
+    if ((node->left == NULL && strcmp(node->value.key, key) > 0) || (node->right == NULL && strcmp(node->value.key, key) < 0)) {
+        return NULL;
+    }
+
+    if (strcmp(node->value.key, key) < 0) {
+        foundValue = searchByKey(node->right, key);
+    }
+    if (strcmp(node->value.key, key) > 0) {
+        foundValue = searchByKey(node->left, key);
+    }
+    return foundValue;
```

Comment:
```suggestion
    if (strcmp(node->value.key, key) < 0) {
        return searchByKey(node->right, key);
    }
    return searchByKey(node->left, key);
```
И тогда foundValue будет вообще не нужно